File: code/reverse_linked_list.java
public class Solution {

    class ListNode {
        int val;
        ListNode next;
        public ListNode(int n) {
            this.val = n;
        }
    }

    /**
     * 4 句话：
     * 1. 新建post
     * 2. 翻转current
     * 3. 后移prev
     * 4. 后移current
     */

    public reverseLinkedList(ListNode head) {

        ListNode prev = null;
        while (head != null) {
            ListNode post = head.next;
            head.next = prev;
            prev = head;
            head = post;
        }
        return prev;

    }


}

File: code/medium1/reverse_linked_list.cpp
#include<iostream>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {

        // ListNode* dummy = new ListNode(0);
        // dummy->next = head;
        ListNode* prev = NULL;
        while (head != NULL) {
            ListNode* post = head->next;
            head->next = prev;
            prev = head;
            head = post;
        }
        return prev;
    }
};